File: FitBook.Model/Responses/UserMemberships/UserMembershipResponse.cs
using FitBook.Model.Enums;

namespace FitBook.Model.Responses.UserMemberships;

public class UserMembershipResponse : IEntityResponse
{
    public int Id { get; set; }

    public MembershipStatus Status { get; set; }

    public DateTime StartDateUtc { get; set; }
    public DateTime EndDateUtc { get; set; }
    public DateTime? NextPaymentDateUtc { get; set; }

    public bool IsActive { get; set; }

    public bool IsPaid { get; set; }

    public int UserAccountId { get; set; }
    public string UserFirstName { get; set; } = string.Empty;
    public string UserLastName { get; set; } = string.Empty;
    public string UserEmail { get; set; } = string.Empty;

    public int MembershipPackageId { get; set; }
    public string PackageName { get; set; } = string.Empty;
    public decimal PackagePrice { get; set; }
    public int PackageDurationDays { get; set; }

    public DateTime CreatedAtUtc { get; set; }
    public DateTime? UpdatedAtUtc { get; set; }
}

File: UI/fitbook_mobile/lib/models/responses/user_membership_response.dart
import 'package:json_annotation/json_annotation.dart';

import '../enums/membership_status.dart';

part 'user_membership_response.g.dart';

@JsonSerializable()
class UserMembershipResponse {
  UserMembershipResponse({
    required this.id,
    required this.status,
    required this.startDateUtc,
    required this.endDateUtc,
    this.nextPaymentDateUtc,
    required this.isActive,
    required this.isPaid,
    required this.userAccountId,
    required this.userFirstName,
    required this.userLastName,
    required this.userEmail,
    required this.membershipPackageId,
    required this.packageName,
    required this.packagePrice,
    required this.packageDurationDays,
    required this.createdAtUtc,
    this.updatedAtUtc,
  });

  final int id;
  final MembershipStatus status;
  final DateTime startDateUtc;
  final DateTime endDateUtc;
  final DateTime? nextPaymentDateUtc;
  final bool isActive;
  final bool isPaid;
  final int userAccountId;
  final String userFirstName;
  final String userLastName;
  final String userEmail;
  final int membershipPackageId;
  final String packageName;
  final double packagePrice;
  final int packageDurationDays;
  final DateTime createdAtUtc;
  final DateTime? updatedAtUtc;

  String get userFullName => '$userFirstName $userLastName';

  factory UserMembershipResponse.fromJson(Map<String, dynamic> json) =>
      _$UserMembershipResponseFromJson(json);

  Map<String, dynamic> toJson() => _$UserMembershipResponseToJson(this);
}
